server/src/main/java/org/elasticsearch/transport/Lz4TransportDecompressor.java

```diff
@@ -0,0 +1,355 @@
+                                decompressor.decompress(compressed, compressedOffset, uncompressed, 0, decompressedLength);
+                                break;
+                            default:
+                                throw new IllegalStateException(String.format(Locale.ROOT,
+                                    "unexpected blockType: %d (expected: %d or %d)",
+                                    blockType, BLOCK_TYPE_NON_COMPRESSED, BLOCK_TYPE_COMPRESSED));
+                        }
+                        // Skip inbound bytes after we processed them.
+                        bytesConsumed += compressedLength;
+
+                        if (checksum != null) {
+                            checksum.reset();
+                            checksum.update(uncompressed, 0, decompressedLength);
+                            final int checksumResult = (int) checksum.getValue();
+                            if (checksumResult != currentChecksum) {
+                                throw new IllegalStateException(String.format(Locale.ROOT,
+                                    "stream corrupted: mismatching checksum: %d (expected: %d)",
+                                    checksumResult, currentChecksum));
+                            }
+                        }
+
+                        int bytesToCopy = decompressedLength;
+                        int uncompressedOffset = 0;
+                        while (bytesToCopy > 0) {
+                            final Recycler.V<byte[]> page;
+                            final boolean isNewPage = pageOffset == PageCacheRecycler.BYTE_PAGE_SIZE;
+                            if (isNewPage) {
+                                pageOffset = 0;
+                                pages.add(recycler.bytePage(false));
+                            }
+                            page = pages.getLast();
+
+                            int toCopy = Math.min(bytesToCopy, PageCacheRecycler.BYTE_PAGE_SIZE - pageOffset);
+                            System.arraycopy(uncompressed, uncompressedOffset, page.v(), pageOffset, toCopy);
+                            pageOffset += toCopy;
+                            bytesToCopy -= toCopy;
+                            uncompressedOffset += toCopy;
+                        }
+                        currentState = State.INIT_BLOCK;
+                    } catch (LZ4Exception e) {
+                        throw new IllegalStateException(e);
+                    }
+                    break;
+                case FINISHED:
+                    break;
+                case CORRUPTED:
+                    throw new IllegalStateException("LZ4 stream corrupted.");
+                default:
+                    throw new IllegalStateException();
+            }
+        } catch (IOException e) {
+            currentState = State.CORRUPTED;
+            throw e;
+        }
+        return bytesConsumed;
+    }
+
+    private byte[] getCompressedBuffer(int requiredSize) {
```

Comment:
The buffer returned here should be 16k at the most right? Maybe we can just pre-allocate it to 16k or make it thread-local to potentially avoid a bunch of allocations for new connections?